numpy/core/src/umath/scalarmath.c.src

```diff
@@ -587,6 +587,9 @@ static NPY_INLINE int
 @name@_ctype_absolute(@type@ a, @type@ *out)
 {
     *out = (a < 0 ? -a : a);
+    if (a < 0 && a == -a) {
```

Comment:
My understanding is that if `a` has a signed integer type and `a` equals the minimum integer, the result of `-a` is *undefined*.  I don't know the best reference for this, but an example where you can find this stated is in Table I on page A:4 of https://www.cs.utah.edu/~regehr/papers/tosem15.pdf.

This code *might* work fine with all the C compilers that we use, but in general, undefined behavior should be avoided.

Perhaps another template loop variable could be added that holds the type-dependent minimum integer name (something from `limits.h` or its numpy equivalent, if numpy has something like it), and a direct equality test could be used.

Reply:
The expresson `a == -a` is also used in `@name@_ctype_divide()`.

Reply:
Fair, maybe we should just use it?  I am sure we have two complement integers hard-coded all over the place and since we will have tests, we should notice if a system comes along where things actually break...

Reply:
I pointed it out thinking we should fix it there, too :grin:.  But I don't know if we want to expand the scope of this PR that much.

Reply:
I incorporated the suggestion to use templating for `NPY_MIN_@NAME@` and can create a follow-up PR to address the same unsafe check in `@name_ctype_divide()` if that would be of interest -- that way we can keep this PR small and easy to review.